File: impl/sdk/java/src/examples/java/io/resonatehq/examples/saga/Saga.java
package io.resonatehq.examples.saga;

import io.resonatehq.resonate.Context;
import io.resonatehq.resonate.Handle.ResonateHandle;
import io.resonatehq.resonate.Resonate;
import io.resonatehq.resonate.Retry.Never;

/**
 * saga shows a multi-step durable workflow with compensation on failure (the canonical "distributed
 * transactions" pattern):
 *
 * <pre>
 *     book_trip:
 *         1. reserve_flight
 *         2. reserve_hotel   (on failure: release_flight)
 *         3. charge_card     (on failure: release_hotel, release_flight)
 * </pre>
 *
 * <p>Each step is its own registered function dispatched via {@code ctx.run}. Step settlement is
 * recorded in a durable promise, so if the worker crashes between two steps a restart skips the steps
 * that already settled and runs only the missing ones -- including the compensations.
 *
 * <p>Start a Resonate server on localhost:8001 first ({@code resonate dev}), then either of:
 *
 * <pre>{@code ./gradlew runExample -PmainClass=io.resonatehq.examples.saga.Saga                       # happy path
 * ./gradlew runExample -PmainClass=io.resonatehq.examples.saga.Saga -PexampleArgs="--fail charge" # both compensations run}</pre>
 *
 * <p>Steps throw <i>ordinary</i> exceptions ({@link BookingError} below) -- there is no need to
 * construct a Resonate {@code ApplicationError}. The orchestrator catches the domain exception by its
 * real type -- the SDK serializes it across the durability boundary and reconstructs it when the
 * awaiting worker can load the class -- with {@code ApplicationError} as the fallback for when it
 * cannot round-trip. Note the catch is deliberately <i>not</i> a bare {@code catch (Throwable)}: that
 * would also swallow the internal {@code Suspended} signal the durable replay machinery relies on.
 *
 * <p>Note on replay: a durable orchestrator re-executes from the top each time it awaits a
 * not-yet-settled future, so any side effect (a {@code println}, an external call) belongs in a leaf
 * step function -- which settles once and never re-runs -- not in {@code bookTrip} itself. That is why
 * every log line below lives in a step, never in the orchestrator.
 */
public final class Saga {
    private Saga() {}

    /**
     * A plain domain error -- deliberately NOT a Resonate error.
     *
     * <p>Demonstrates that step functions can throw any exception; the SDK settles the step's promise
     * {@code rejected} and reconstructs the original type for the orchestrator when it can round-trip
     * across the durability boundary (same runtime, loadable class), falling back to {@code
     * ApplicationError} otherwise.
     */
    public static class BookingError extends RuntimeException {
        private static final long serialVersionUID = 1L;

        public BookingError(String message) {
            super(message);
        }
    }

    public static final class HotelReservationError extends BookingError {
        private static final long serialVersionUID = 1L;

        public HotelReservationError(String message) {
            super(message);
        }
    }

    public static final class ChargeCardError extends BookingError {
        private static final long serialVersionUID = 1L;

        public ChargeCardError(String message) {
            super(message);
        }
    }

    /** The triple returned by {@code bookTrip}. */
    public record TripResult(String flight, String hotel, String charge) {}

    // -- Step functions (leaves: each prints once, settles once) ---------------

    public static String reserveFlight(Context ctx, String customer, String frm, String to) {
        String ref = "FL-" + customer + "-" + frm + "-" + to;
        System.out.println("  [reserve_flight] reserved " + ref);
        return ref;
    }

    public static String reserveHotel(Context ctx, String customer, String city, boolean fail) {
        if (fail) {
            System.out.println("  [reserve_hotel] FAILED for " + customer + " in " + city);
            throw new HotelReservationError("no rooms available in " + city);
        }
        String ref = "HT-" + customer + "-" + city;
        System.out.println("  [reserve_hotel] reserved " + ref);
        return ref;
    }

    public static String chargeCard(Context ctx, String customer, int amount, boolean fail) {
        if (fail) {
            System.out.println("  [charge_card] FAILED for " + customer + " ($" + amount + ")");
            throw new ChargeCardError("card declined for $" + amount);
        }
        String ref = "CH-" + customer + "-" + amount;
        System.out.println("  [charge_card] charged " + ref);
        return ref;
    }

    public static String releaseFlight(Context ctx, String ref) {
        System.out.println("  [release_flight] released " + ref);
        return ref;
    }

    public static String releaseHotel(Context ctx, String ref) {
        System.out.println("  [release_hotel] released " + ref);
        return ref;
    }

    // -- Saga orchestrator -----------------------------------------------------

    public static TripResult bookTrip(Context ctx, String customer, String frm, String to, int amount, String failAt) {
        // Step 1: flight
        String flight = ctx.run(Saga::reserveFlight, customer, frm, to).await();

        // Step 2: hotel (compensate the flight on failure)
        String hotel;
        try {
            hotel = ctx.run(Saga::reserveHotel, customer, to, "hotel".equals(failAt))
                    .await();
        } catch (BookingError e) {
            compensate(ctx, "", flight);
            throw e;
        }

        // Step 3: charge (compensate hotel + flight on failure, reverse order)
        String charge;
        try {
            charge = ctx.run(Saga::chargeCard, customer, amount, "charge".equals(failAt))
                    .await();
        } catch (BookingError e) {
            compensate(ctx, hotel, flight);
            throw e;
        }

        return new TripResult(flight, hotel, charge);
    }

    /**
     * Run the inverse of any completed steps in reverse order. Empty refs are skipped. Each
     * compensation is its own durable promise, so a crash mid-rollback resumes at the first unsettled
     * one.
     */
    private static void compensate(Context ctx, String hotelRef, String flightRef) {
        if (!hotelRef.isEmpty()) {
            ctx.run(Saga::releaseHotel, hotelRef).await();
        }
        if (!flightRef.isEmpty()) {
            ctx.run(Saga::releaseFlight, flightRef).await();
        }
    }

    public static void main(String[] args) {
        String fail = argValue(args, "--fail", "");

        String url = System.getenv().getOrDefault("RESONATE_URL", "http://localhost:8001");
        Resonate r = Resonate.builder().url(url).retryPolicy(new Never()).build();
        r.register(Saga::bookTrip);
        r.register(Saga::reserveFlight);
        r.register(Saga::reserveHotel);
        r.register(Saga::chargeCard);
        r.register(Saga::releaseFlight);
        r.register(Saga::releaseHotel);

        try {
            String id = "saga-" + System.nanoTime();
            System.out.println("[book_trip] starting workflow id=" + id + " fail_at='" + fail + "'");
            ResonateHandle<TripResult> handle = r.run(id, Saga::bookTrip, "alice", "SFO", "JFK", 850, fail);
            TripResult trip;
            try {
                trip = handle.result();
            } catch (BookingError exc) {
                System.out.println("[book_trip] FAILED: " + exc.getClass().getSimpleName() + ": " + exc.getMessage());
                return;
            }
            System.out.printf(
                    "[book_trip] OK: flight=%s hotel=%s charge=%s%n", trip.flight(), trip.hotel(), trip.charge());
            assert "FL-alice-SFO-JFK".equals(trip.flight());
            assert "HT-alice-JFK".equals(trip.hotel());
            assert "CH-alice-850".equals(trip.charge());
        } finally {
            r.stop();
        }
    }

    private static String argValue(String[] args, String flag, String fallback) {
        for (int i = 0; i + 1 < args.length; i++) {
            if (args[i].equals(flag)) {
                return args[i + 1];
            }
        }
        return fallback;
    }
}

File: impl/sdk/ts/examples/generator/saga.ts
// saga shows a multi-step durable workflow with compensation on failure
// (the canonical "distributed transactions" pattern):
//
//   bookTrip:
//     1. reserveFlight
//     2. reserveHotel   (on failure: releaseFlight)
//     3. chargeCard     (on failure: releaseHotel, releaseFlight)
//
// Each step is its own registered function dispatched via ctx.rpc. Step
// settlement is recorded in a durable promise, so if the worker crashes between
// two steps a restart skips the settled steps and runs only the missing ones --
// including the compensations. A failing step rejects, which we catch to drive
// the rollback.
//
// Start a Resonate server on localhost:8001 first (`resonate dev`), then:
//
//   npx tsx examples/generator/saga.ts                # happy path
//   npx tsx examples/generator/saga.ts --fail charge  # forces chargeCard to
//                                                     # fail; both compensations
//                                                     # run in reverse order

import { type Context, Never, Resonate } from "../../src/index.js";

type FailAt = "" | "hotel" | "charge";

interface TripResult {
  flightRef: string;
  hotelRef: string;
  chargeRef: string;
}

// -- Step functions (leaves: each performs one action) --------------------

async function reserveFlight(_ctx: Context, customer: string, from: string, to: string): Promise<string> {
  const ref = `FL-${customer}-${from}-${to}`;
  console.log(`  [reserveFlight] reserved ${ref}`);
  return ref;
}

async function reserveHotel(_ctx: Context, customer: string, city: string, fail: boolean): Promise<string> {
  if (fail) {
    console.log(`  [reserveHotel] FAILED for ${customer} in ${city}`);
    throw new Error(`no rooms available in ${city}`);
  }
  const ref = `HT-${customer}-${city}`;
  console.log(`  [reserveHotel] reserved ${ref}`);
  return ref;
}

async function chargeCard(_ctx: Context, customer: string, amount: number, fail: boolean): Promise<string> {
  if (fail) {
    console.log(`  [chargeCard] FAILED for ${customer} ($${amount})`);
    throw new Error(`card declined for $${amount}`);
  }
  const ref = `CH-${customer}-${amount}`;
  console.log(`  [chargeCard] charged ${ref}`);
  return ref;
}

async function releaseFlight(_ctx: Context, ref: string): Promise<string> {
  console.log(`  [releaseFlight] released ${ref}`);
  return ref;
}

async function releaseHotel(_ctx: Context, ref: string): Promise<string> {
  console.log(`  [releaseHotel] released ${ref}`);
  return ref;
}

// -- Saga orchestrator ----------------------------------------------------

function* bookTrip(
  ctx: Context,
  customer: string,
  from: string,
  to: string,
  amount: number,
  failAt: FailAt,
): Generator<any, TripResult, any> {
  // Step 1: flight
  const flightRef: string = yield* ctx.rpc<string>("reserveFlight", customer, from, to);

  // Step 2: hotel -- on failure, release the flight. Never so a forced failure
  // surfaces at once instead of retrying the leaf.
  const noRetry = ctx.options({ retryPolicy: new Never() });
  let hotelRef: string;
  try {
    hotelRef = yield* ctx.rpc<string>("reserveHotel", customer, to, failAt === "hotel", noRetry);
  } catch (err) {
    yield* compensate(ctx, "", flightRef);
    throw new Error(`reserveHotel: ${(err as Error).message}`);
  }

  // Step 3: charge -- on failure, release hotel then flight
  let chargeRef: string;
  try {
    chargeRef = yield* ctx.rpc<string>("chargeCard", customer, amount, failAt === "charge", noRetry);
  } catch (err) {
    yield* compensate(ctx, hotelRef, flightRef);
    throw new Error(`chargeCard: ${(err as Error).message}`);
  }

  return { flightRef, hotelRef, chargeRef };
}

// compensate runs the inverse of completed steps in reverse order. Empty refs
// are skipped. Compensation failures are logged, not surfaced -- the saga has
// already failed; the goal is best-effort rollback.
function* compensate(ctx: Context, hotelRef: string, flightRef: string): Generator<any, void, any> {
  console.log("  [bookTrip] running compensations...");
  if (hotelRef) {
    try {
      yield* ctx.rpc<string>("releaseHotel", hotelRef);
    } catch (err) {
      console.log(`  [bookTrip] releaseHotel failed: ${(err as Error).message}`);
    }
  }
  if (flightRef) {
    try {
      yield* ctx.rpc<string>("releaseFlight", flightRef);
    } catch (err) {
      console.log(`  [bookTrip] releaseFlight failed: ${(err as Error).message}`);
    }
  }
}

// -- main -----------------------------------------------------------------

function flag(name: string, fallback: string): string {
  const i = process.argv.indexOf(`--${name}`);
  return i >= 0 && process.argv[i + 1] ? process.argv[i + 1] : fallback;
}

const failAt = flag("fail", "") as FailAt;

const resonate = new Resonate({ url: process.env.RESONATE_URL ?? "http://localhost:8001" });
resonate.register("bookTrip", bookTrip);
resonate.register("reserveFlight", reserveFlight);
resonate.register("reserveHotel", reserveHotel);
resonate.register("chargeCard", chargeCard);
resonate.register("releaseFlight", releaseFlight);
resonate.register("releaseHotel", releaseHotel);

const id = `saga-${Date.now()}`;
console.log(`[bookTrip] starting workflow id=${id} fail_at=${JSON.stringify(failAt)}`);

try {
  const out: TripResult = await resonate.run(id, bookTrip, "alice", "SFO", "JFK", 850, failAt);
  console.log(`[bookTrip] OK: flight=${out.flightRef} hotel=${out.hotelRef} charge=${out.chargeRef}`);
} catch (err) {
  console.log(`[bookTrip] FAILED: ${(err as Error).message}`);
} finally {
  await resonate.stop();
}
